### DISCUSSION TITLE
hledger support

### DISCUSSION THREAD

**Discussion by simonmichael** [ananthakumaran/paisa#12] [Ideas]
Paisa looks outstanding! Congratulations and thank you for sharing it.

I was able to make my hledger journal Ledger-readable so I could view it in Paisa, but it would be much more convenient if Paisa could read it directly. Would love to see optional support for reading with hledger. I started some related notes at https://hledger.org/paisa.html#tips .

**Comment by ananthakumaran** [comment]
Thanks for checking out paisa. Regarding hledger support, I will look into it but it will take some time since I am not familiar with the difference between ledger and hledger. 

There are a few known issues with the way I use the ledger now

1) I couldn't determine how to get the unit price via CSV export, so I derived it by performing `amount / quantity`. This would lead to rounding issues and currently working around it by scaling up amount `%(quoted(to_int(scrub(market(amount,date) * 100000))))`
2) I have plans to add support for editing journal files from browser. I tried to use code mirror, but that required a lot of work on the parser side, so I paused the work
3) [`"`](https://github.com/ledger/ledger/issues/2007) are not handled properly by ledger export, I am working around that as well
4) depending on binary is cumbersome, even though the above got fixed recently, I can't remove the workaround, since the version is too new and might not be available in all distros

I will eventually come back to this and I might have to write my own parser for ledger files.

**Reply by edkedk99** [reply]
I was also impressed by you abilities in web development and financial analysis. 

I recently faced a similar problem as your item "1" but using hledger. I could solve it by parsing the json export instead of the csv. The "csv" output in hledger gives us subset of transaction data, while json output in hledger gives all the data, including if it has price and quantity or total amount. It even give the position of the transaction in the journal file.

I am not sure if ledger export is limited to csv only.

**Comment by simonmichael** [comment]
Thanks for your interest @ananthakumaran . 

For now the differences would come down to running `hledger -f FILE print -O csv` (and interpreting the different column layout) instead of `ledger -f FILE csv --csv-format ...`. This could be done automatically if `hledger` is found in PATH. Or, it could be possible to configure a command in the config file, that is responsible for producing CSV in a paisa friendly format..

It was interesting to hear about the issues. I guess you probably will write your own parser eventually, but that will always work for only a subset of user files - it will "never" reach 100% compatibility - so I think it's cool that you support using the original tool at least as a fallback, because that can be relied on to parse its own files.

Editing is an interesting topic. I like your current model of repeatedly updating one way from journal file to sqlite. I guess you could easily write back to sqlite, but that would be hard to sync back to the journal file (or files, there could be many). My hledger-web app reads from the journal files directly, with no sqlite db. It can't write back to the journal either (except by appending, or by replacing the whole thing). But it knows the textual position of each transaction's journal entry, so in principle it could edit individual transactions. I suppose you could maintain similar info through the sqlite db as well.

**Reply by ananthakumaran** [reply]
I am trying to build a code editor to edit the ledger files directly [poc](https://paisa-demo.ananthakumaran.in/ledger/editor). This means I don't have to worry about parsing, as I can leave it to the ledger. Yet to figure out whether this is usable or if the majority prefer form based GUI for CRUD.

**Comment by ananthakumaran** [comment]
hledger support is available on 0.3.6

**Comment by simonmichael** [comment]
Thank you very much! I look forward to trying it out.

**Comment by nickdex** [comment]
![Screenshot 2023-10-0
![Screenshot 2023-10-09 at 7 00 12 PM](https://github.com/ananthakumaran/paisa/assets/8512967/6adefcfc-1db9-4d75-89ef-e3229bbc6037)
9 at 6 59 39 PM](https://github.com/ananthakumaran/paisa/assets/8512967/a8d0322b-304b-4e52-8c5c-bb1d49f8b4da)
![Screenshot 2023-10-09 at 7 01 08 PM](https://github.com/ananthakumaran/paisa/assets/8512967/0bd6d68f-91ff-4c4c-8897-6dd6ca1c967a)
![Screenshot 2023-10-09 at 7 02 26 PM](https://github.com/ananthakumaran/paisa/assets/8512967/cfaf0e3e-ef4f-463c-a08c-0967712c85f5)
I have hledger installed (using homebrew) and have it $PATH as well (checked both bash and zsh shells). Posted screenshots for path and running hledger

But when I choose hledger in paisa desktop app it says 
![Screenshot 2023-10-09 at 7 04 32 PM](https://github.com/ananthakumaran/paisa/assets/8512967/889c58c3-a3d9-40cd-b50e-7767692322a4)
`hledger not found in $PATH` @ananthakumaran  any idea how to fix it?

**Reply by ananthakumaran** [reply]
This is Mac desktop app? desktop app doesn't seem to set the `PATH` env variable. Paisa searches in few known folders, if you copy hledger to one of those it will work.

https://github.com/ananthakumaran/paisa/blob/f02323ca62669ecad2393d97d9a174a76bc58f33/internal/binary/embed.go#L14-L20

**Reply by nickdex** [reply]
@ananthakumaran It worked! Thanks

**Comment by n1zzo** [comment]
It could be added to the Docker image as well, here's a PR that fixes that:

https://github.com/ananthakumaran/paisa/pull/113

**Reply by ananthakumaran** [reply]
I don't have plans to ship hledger/beancount with paisa, I am expecting the end user to install it. For docker, you could start from paisa as base image and add hledger.

**Reply by f-izzo** [reply]
I wanted to use paisa with hledger too.
Here is the Dockerfile using paisa as base image, as @ananthakumaran suggested.

Dockerfile:
```dockerfile
# syntax=docker/dockerfile:1
FROM ananthakumaran/paisa:latest
RUN apk --no-cache add hledger
```
To use this, place the Dockerfile in the same folder as your `docker-compose.yaml` and 
replace `image: ananthakumaran/paisa:latest` with `build: .` in your `docker-compose.yaml`.
